File: Python/utils/crud_operations.py
import psycopg2
from typing import Optional, Union


class CrudOperations:
    @staticmethod
    def connect_to_db(db_name: str, user: str, password: str) -> psycopg2.extensions.connection:
        conn: Optional[psycopg2.extensions.connection] = None
        try:
            conn = psycopg2.connect(
                host="localhost",
                port=5432,
                database=db_name,
                user=user,
                password=password
            )
        except (Exception, psycopg2.Error) as error:
            print(f"Error while connecting to the PostgreSQL server: {error}")
        finally:
            return conn

    @staticmethod
    def close_connection(con: psycopg2.extensions.connection) -> None:
        if con:
            try:
                con.close()
            except (Exception, psycopg2.Error) as error:
                print(f"Error while closing the database connection: {error}")

    @staticmethod
    def execute_query(con: psycopg2.extensions.connection, query: str) -> Union[list, int]:
        cur = None
        try:
            cur = con.cursor()
            cur.execute(query)
            if cur.description:  # SELECT query
                rows = cur.fetchall()
            elif cur.rowcount == -1:  # CREATE TABLE
                rows = 0
            else:  # INSERT, UPDATE, DELETE queries
                rows = cur.rowcount
        except (Exception, psycopg2.DatabaseError) as error:
            print(f"Error: {error}")
            rows = -1
        finally:
            if cur:
                cur.close()
        return rows

    @staticmethod
    def print_query_result(queryStr: str, file_path: str, rows: Union[list, int]) -> None:
        with open(file_path, "a") as f:
            print(f"{queryStr}", file=f)
            if isinstance(rows, list):
                for row in rows:
                    print("|".join(map(str, row)), file=f)
                print("\n\n\n\n", file=f)
            else:
                if rows != -1:
                    print(f"{rows} rows affected.", file=f)
                    print("\n\n\n\n", file=f)

File: Go/utils/crud_operations.go
package utils

import (
	"context"
	"fmt"
	"os"

	"github.com/jackc/pgx/v5"
)

func ConnectToDB(dbName, user, password, host, port string) (*pgx.Conn, error) {
	url := fmt.Sprintf("postgres://%s:%s@%s:%s/%s", user, password, host, port, dbName)
	return pgx.Connect(context.Background(), url)
}

func CloseDB(conn *pgx.Conn) {
	conn.Close(context.Background())
}

func ExecuteQuery(conn *pgx.Conn, query string) (pgx.Rows, error) {
	return conn.Query(context.Background(), query)
}

func ExecuteUpdateQuery(conn *pgx.Conn, query string) (int, error) {
	ct, err := conn.Exec(context.Background(), query)
	if err != nil {
		return -1, err
	}
	return int(ct.RowsAffected()), nil
}

func PrintQueryResultToFile(queryStr string, rows pgx.Rows, filePath string) error {
	file, err := os.OpenFile(filePath, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0644)
	if err != nil {
		return err
	}
	defer file.Close()

	_, err = file.WriteString(queryStr + ":\n")
	if err != nil {
		return err
	}

	for rows.Next() {
		values, err := rows.Values()
		if err != nil {
			return err
		}
		resultString := ""
		for _, value := range values {
			resultString += fmt.Sprintf("%v", value) + "|"
		}
		resultString = resultString[:len(resultString)-1]

		_, err = file.WriteString(resultString + "\n")
		if err != nil {
			return err
		}
	}
	_, err = file.WriteString("\n\n\n\n\n")
	if err != nil {
		return err
	}
	return nil
}

func PrintRowsAffectedToFile(queryStr string, rowsAffected int, filePath string) error {
	file, err := os.OpenFile(filePath, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0644)
	if err != nil {
		return err
	}
	defer file.Close()

	_, err = file.WriteString(queryStr + ":\n")
	if err != nil {
		return err
	}

	_, err = file.WriteString(fmt.Sprintf("Rows Affected: %d\n\n\n\n\n\n", rowsAffected))
	if err != nil {
		return err
	}
	return nil
}